server/src/session.ts

```diff
@@ -197,6 +200,59 @@ export class Session {
     conn.onCodeLens(p => this.onCodeLens(p));
     conn.onCodeLensResolve(p => this.onCodeLensResolve(p));
     conn.onSignatureHelp(p => this.onSignatureHelp(p));
+    conn.onCodeAction(p => this.codeAction(p));
+    conn.onCodeActionResolve(p => this.onCodeActionResolve(p));
+  }
+
+  private codeAction(params: lsp.CodeActionParams): lsp.CodeAction[]|null {
+    const filePath = uriToFilePath(params.textDocument.uri);
+    const lsInfo = this.getLSAndScriptInfo(params.textDocument);
+    if (!lsInfo) {
+      return null;
+    }
+    const start = lspPositionToTsPosition(lsInfo.scriptInfo, params.range.start);
+    const end = lspPositionToTsPosition(lsInfo.scriptInfo, params.range.end);
+    const errorCodes = params.context.diagnostics.map(diag => diag.code)
+                           .filter((code): code is number => typeof code === 'number');
+
+    const codeActions = lsInfo.languageService.getCodeFixesAtPosition(
+        filePath, start, end, errorCodes, defaultFormatOptions, defaultPreferences);
+    return codeActions
+        .map<lsp.CodeAction>(codeAction => {
+          return {
+            title: codeAction.description,
+            kind: lsp.CodeActionKind.QuickFix,
+            diagnostics: params.context.diagnostics,
+            edit: tsFileTextChangesToLspWorkspaceEdit(
+                codeAction.changes, (path: string) => this.projectService.getScriptInfo(path)),
+          };
+        })
+        .concat(getCodeFixesAll(codeActions, params.textDocument));
+  }
+
+  private onCodeActionResolve(param: lsp.CodeAction): lsp.CodeAction {
+    const codeActionResolve = param.data as unknown as CodeActionResolveData;
+    const isCodeFixesAll = codeActionResolve.fixId !== undefined;
+    if (!isCodeFixesAll) {
+      return param;
+    }
```

Comment:
Can you add a quick comment about why we don't need to do any resolving for things that aren't "codeFixesAll"? And a comment/link that describes why the presence or lack of a `fixId` is the distinguishing character for this type of fix.